### ISSUE TITLE
crash after calling collectDeviceData, after in-app PayPal authorisation

### ISSUE THREAD

**Issue by jamesrgrinter** [braintree/braintree_ios#108]
Per the new requirements of vaulted PayPal payment methods, we're using the BTData mechanism to collect device data.

I'm invoking this in our dropInViewController:didSucceedWithPaymentMethod: delegate (MWBTDataEnvironment being BTDataEnvironmentSandbox for my development build, and BTDataEnvironmentProduction for my live one)

    BTData *bTData = [[BTData alloc] initWithClient:viewController.client environment:MWBTDataEnvironment];
    NSString *deviceData = [bTData collectDeviceData];

immediately before dismissing the dropInViewController.

    [self dismissViewControllerAnimated:YES completion:nil];

Normally this works, but in one particular case - if the user is adding a new payment method of PayPal, and they're doing it in-app (rather than having the PayPal app installed and bouncing in and out of that) - we'll get a crash.

Crashlytics catches this as:

    Thread : Crashed: com.apple.main-thread
    0  libobjc.A.dylib                0x38d27626 objc_msgSend + 5
    1  UIKit                          0x30faa3f1 -[UIWebView webView:decidePolicyForNavigationAction:request:frame:decisionListener:] + 184
    2  CoreFoundation                 0x2e570534 __invoking___ + 68
    3  CoreFoundation                 0x2e4bb08b -[NSInvocation invoke] + 282
    4  CoreFoundation                 0x2e4be823 -[NSInvocation invokeWithTarget:] + 50
    5  WebKit                         0x37128b33 -[_WebSafeForwarder forwardInvocation:] + 222
    6  CoreFoundation                 0x2e56f0a1 ___forwarding___ + 352
    7  CoreFoundation                 0x2e4be768 _CF_forwarding_prep_0 + 24
    8  CoreFoundation                 0x2e570534 __invoking___ + 68
    9  CoreFoundation                 0x2e4bb08b -[NSInvocation invoke] + 282
    10 WebCore                        0x36783035 HandleDelegateSource(void*) + 100
    11 CoreFoundation                 0x2e53820b __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 14
    12 CoreFoundation                 0x2e5376db __CFRunLoopDoSources0 + 206
    13 CoreFoundation                 0x2e535ecf __CFRunLoopRun + 622
    14 CoreFoundation                 0x2e4a0ebf CFRunLoopRunSpecific + 522
    15 CoreFoundation                 0x2e4a0ca3 CFRunLoopRunInMode + 106
    16 GraphicsServices               0x333a6663 GSEventRunModal + 138
    17 UIKit                          0x30ded14d UIApplicationMain + 1136
    18 My app                         0x0017f029 main (main.m:16)

and, I get what I suspect is a related message (this, from an earlier crash when running from Xcode, so any relevant addresses will not match)

    [DeviceCollectorSDK respondsToSelector:]: message sent to deallocated instance 0x1e50e880

Now I know that it's not crashing immediately, because the device data is being correctly submitted to my own server. Also, it does not crash when using that payment method next time around, so it seems to be connected to the in-app PayPal authorisation.

Thoughts about where to look, or what it could be?

**Comment by mickeyreiss** [comment]
Given this error message (`message sent to deallocated instance`), I'm wondering if you could workaround this problem by retaining your instance of `BTData` in a property.

**Comment by mickeyreiss** [comment]
Any update on whether this workaround helepd?

**Comment by jamesrgrinter** [comment]
I have to do a bit more testing but yes, I think it has.

**Comment by burnto** [comment]
We believe this is now resolved, but @jamesrgrinter we'll re-open if you reply here stating otherwise!

**Comment by Reflejo** [comment]
This is an issue for us and retaining `BTData` doesn't work for us. Any chance this can be reopened? This issue happens because a missing `webView.delegate = nil` from `DeviceCollectorSDK`

cc @mickeyreiss

**Comment by Reflejo** [comment]
Any news on this?. We have a nasty hack to make this work.

**Comment by burnto** [comment]
@skunkworks does this still apply in 4.0?

**Comment by skunkworks** [comment]
@Reflejo, I'd like to better understand the problem so I can unblock you.

Is this still causing you problems? If so, have you tried out v4's `BTDataCollector`? Please feel free to open a new issue and I'll do my best to help.

**Comment by Reflejo** [comment]
@skunkworks The problem is(was?) that the `UIWebView` delegate wasn't being `nil` out on `dealloc` causing race conditions as the case above.

We had to add this beauty:

extension BTDataCollector {

```swift
    func generateChargeToken() -> String {
        let chargeToken = self.collectFraudData()
        let kount = self.valueForKey("kount") as? NSObject
        let webView = kount?.valueForKey("redirectWebview") as? UIWebView
        webView?.delegate = nil
        return chargeToken
    }
}```

**Comment by skunkworks** [comment]
@Reflejo Oof, nice hack. Let me do some investigation and reach out to Kount if necessary.

**Comment by skunkworks** [comment]
@Reflejo, Kount has reached out to ask a few questions:

1. Without your workaround, is the issue consistent or intermittent? I'm assuming it's intermittent, since you call it a race condition.
2. Your app is using ARC, correct?
3. Is it possible to get a stack trace for this crash?

**Comment by Reflejo** [comment]
@skunkworks 

1. Doesn't happen every time but does happen quite often. 
2. Full swift app, so yes: ARC.
3. I remember it was the same stacktrace as the one posted in this issue

**Comment by skunkworks** [comment]
@Reflejo We've released version `4.1.1` of the Braintree iOS SDK, which has the latest version of Kount. Is it possible to try this version out to see if it still has the same problem?

**Comment by intelliot** [comment]
For what it's worth -- according to Kount, their library already sets the UIWebView delegate to nil in the dealloc method. This has been true since at least Braintree iOS 3.9.4 (2015-09-25) which uses DeviceCollectorSDK 2.6.0.

So, it's very possible that this issue was fixed as of 3.9.4. Would be good to confirm this :)

**Comment by Reflejo** [comment]
If you guys can confirm this I'll make the change. The thing is that we saw this problem in production and I want to make sure this is fixed before removing the hack.

**Comment by Reflejo** [comment]
So far so good, I think we can close this issue

**Comment by skunkworks** [comment]
Thanks for reporting back :)
